src/hotspot/cpu/s390/gc/shared/barrierSetNMethod_s390.cpp

```diff
@@ -23,18 +23,103 @@
  */
 
 #include "precompiled.hpp"
+#include "asm/assembler.inline.hpp"
+#include "code/codeBlob.hpp"
+#include "code/nativeInst.hpp"
+#include "code/nmethod.hpp"
 #include "gc/shared/barrierSetNMethod.hpp"
 #include "utilities/debug.hpp"
 
+class NativeMethodBarrier: public NativeInstruction {
+  private:
+    static const int GUARD_INSTRUCTION_OFFSET = 3*6; // bytes
+
+    address get_barrier_start_address() const {
+      return NativeInstruction::addr_at(0);
+    }
+
+    NativeMovRegMem* get_patchable_instruction_handle() const {
+      address guard_addr = get_barrier_start_address() + GUARD_INSTRUCTION_OFFSET;
+
+      return reinterpret_cast<NativeMovRegMem*>(guard_addr);
+    }
+
+  public:
+    static const int BARRIER_TOTAL_LENGTH = GUARD_INSTRUCTION_OFFSET + 2*6 + 2; // bytes
+
+    int get_guard_value() const {
+      NativeMovRegMem* guard_addr = get_patchable_instruction_handle();
+
+      // Return guard instruction and value
+      return guard_addr->offset();
+    }
+
+    void set_guard_value(int value) {
+      NativeMovRegMem* guard_addr = get_patchable_instruction_handle();
+
+      // Set guard instruction and value
+      guard_addr->set_offset(value);
+    }
+
+    void verify() const {
```

Comment:
Shouldn't the complete function body be encapsulated with 
#ifdef ASSERT
If ASSERT is undefined, the code has no externally visible effect and thus is irrelevant.

Reply:
This change is reasonable. Thanks for the suggestion.